File: java/tests/com/google/privacy/differentialprivacy/CountConfidenceIntervalTest.java
//
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package com.google.privacy.differentialprivacy;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests the confidence intervals provided by {@link Count}. */
@RunWith(JUnit4.class)
public class CountConfidenceIntervalTest {
  private static final double ARBITRARY_EPSILON = 0.5;
  private static final double ARBITRARY_DELTA = 0.00001;
  private static final int ARBITRARY_MAX_PARTITIONS_CONTRIBUTED = 1;
  private static final double ARBITRARY_ALPHA = 0.23645;

  private Count.Params.Builder builder;

  @Before
  public void setUp() {
    builder =
        Count.builder()
            .epsilon(ARBITRARY_EPSILON)
            .delta(ARBITRARY_DELTA)
            .noise(new GaussianNoise())
            .maxPartitionsContributed(ARBITRARY_MAX_PARTITIONS_CONTRIBUTED);
  }

  @Test
  public void computeConfidenceInterval_clampsNegativeSubinterval() {
    for (int i = 0; i < 1000; i++) {
      Count count = builder.build();
      count.computeResult();

      // Using a large alpha to get a small confidence interval. This increases the chance of both
      // the lower and the upper bound being clamped.
      assertThat(count.computeConfidenceInterval(/*alpha=*/ 0.99).lowerBound()).isAtLeast(0.0);
      assertThat(count.computeConfidenceInterval(/*alpha=*/ 0.99).upperBound()).isAtLeast(0.0);

      // Using a small alpha to get a large confidence interval. This increases the chance of only
      // the the upper bound being clamped.
      assertThat(count.computeConfidenceInterval(/*alpha=*/ 0.01).lowerBound()).isAtLeast(0.0);
      assertThat(count.computeConfidenceInterval(/*alpha=*/ 0.01).upperBound()).isAtLeast(0.0);
    }
  }

  @Test
  public void computeConfidenceInterval_gaussianNoise_calledTwiceForSameAlpha_returnsSameResult() {
    Count count = builder.noise(new GaussianNoise()).build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA))
        .isEqualTo(count.computeConfidenceInterval(ARBITRARY_ALPHA));
  }

  @Test
  public void computeConfidenceInterval_laplaceNoise_calledTwiceForSameAlpha_returnsSameResult() {
    Count count = builder.noise(new LaplaceNoise()).delta(null).build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA))
        .isEqualTo(count.computeConfidenceInterval(ARBITRARY_ALPHA));
  }

  @Test
  public void
      computeConfidenceInterval_gaussianNoise_resultForSmallAlphaContainedInResultForLargeAlpha() {
    Count count = builder.noise(new GaussianNoise()).build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA * 0.5).lowerBound())
        .isLessThan(count.computeConfidenceInterval(ARBITRARY_ALPHA).lowerBound());
    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA * 0.5).upperBound())
        .isGreaterThan(count.computeConfidenceInterval(ARBITRARY_ALPHA).upperBound());
  }

  @Test
  public void
      computeConfidenceInterval_laplaceNoise_resultForSmallAlphaContainedInResultForLargeAlpha() {
    Count count = builder.noise(new LaplaceNoise()).delta(null).build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA * 0.5).lowerBound())
        .isLessThan(count.computeConfidenceInterval(ARBITRARY_ALPHA).lowerBound());
    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA * 0.5).upperBound())
        .isGreaterThan(count.computeConfidenceInterval(ARBITRARY_ALPHA).upperBound());
  }

  @Test
  public void
      computeConfidenceInterval_gaussianNoise_resultMatchesConfidenceIntervalOfGaussianNoise() {
    Count count =
        builder
            .epsilon(ARBITRARY_EPSILON)
            .delta(ARBITRARY_DELTA)
            .noise(new GaussianNoise())
            .maxPartitionsContributed(ARBITRARY_MAX_PARTITIONS_CONTRIBUTED)
            .build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    long result = count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA))
        .isEqualTo(
            new GaussianNoise()
                .computeConfidenceInterval(
                    result,
                    /*l0Sensitivity=*/ ARBITRARY_MAX_PARTITIONS_CONTRIBUTED,
                    /*lInfSensitivity=*/ 1,
                    ARBITRARY_EPSILON,
                    ARBITRARY_DELTA,
                    ARBITRARY_ALPHA));
  }

  @Test
  public void
      computeConfidenceInterval_laplaceNoise_resultMatchesConfidenceIntervalOfLaplaceNoise() {
    Count count =
        builder
            .epsilon(ARBITRARY_EPSILON)
            .delta(null)
            .noise(new LaplaceNoise())
            .maxPartitionsContributed(ARBITRARY_MAX_PARTITIONS_CONTRIBUTED)
            .build();
    count.incrementBy(100000000); // incrementing by large number to prevent clamping
    long result = count.computeResult();

    assertThat(count.computeConfidenceInterval(ARBITRARY_ALPHA))
        .isEqualTo(
            new LaplaceNoise()
                .computeConfidenceInterval(
                    result,
                    /*l0Sensitivity=*/ ARBITRARY_MAX_PARTITIONS_CONTRIBUTED,
                    /*lInfSensitivity=*/ 1,
                    ARBITRARY_EPSILON,
                    /*delta=*/ null,
                    ARBITRARY_ALPHA));
  }

  @Test
  public void computeConfidenceInterval_gaussianNoise_smallAlpha_satisfiesConfidenceLevel() {
    builder.noise(new GaussianNoise());
    int rawCount = 14523;

    int hits = 0;
    for (int i = 0; i < 100000; i++) {
      Count count = builder.build();
      count.incrementBy(rawCount);
      count.computeResult();

      if (count.computeConfidenceInterval(/*alpha=*/ 0.1).lowerBound() <= rawCount
          && rawCount <= count.computeConfidenceInterval(/*alpha=*/ 0.1).upperBound()) {
        hits++;
      }
    }
    // Assuming that the true alpha of the confidence interval mechanism is 0.1, i.e., the raw value
    // is within the confidence interval with probability of at least 0.9, then the hits count will
    // be at least 89546 with probability greater than 1 - 10^-6.
    assertThat(hits).isAtLeast(89546);
  }

  @Test
  public void computeConfidenceInterval_laplaceNoise_smallAlpha_satisfiesConfidenceLevel() {
    builder.noise(new LaplaceNoise()).delta(0.0);
    int rawCount = 14523;

    int hits = 0;
    for (int i = 0; i < 100000; i++) {
      Count count = builder.build();
      count.incrementBy(rawCount);
      count.computeResult();

      if (count.computeConfidenceInterval(/*alpha=*/ 0.1).lowerBound() <= rawCount
          && rawCount <= count.computeConfidenceInterval(/*alpha=*/ 0.1).upperBound()) {
        hits++;
      }
    }
    // Assuming that the true alpha of the confidence interval mechanism is 0.1, i.e., the raw value
    // is within the confidence interval with probability of at least 0.9, then the hits count will
    // be at least 89546 with probability greater than 1 - 10^-6.
    assertThat(hits).isAtLeast(89546);
  }

  @Test
  public void computeConfidenceInterval_gaussianNoise_largeAlpha_satisfiesConfidenceLevel() {
    builder.noise(new GaussianNoise());
    int rawCount = 14523;

    int hits = 0;
    for (int i = 0; i < 100000; i++) {
      Count count = builder.build();
      count.incrementBy(rawCount);
      count.computeResult();

      if (count.computeConfidenceInterval(/*alpha=*/ 0.9).lowerBound() <= rawCount
          && rawCount <= count.computeConfidenceInterval(/*alpha=*/ 0.9).upperBound()) {
        hits++;
      }
    }
    // Assuming that the true alpha of the confidence interval mechanism is 0.9, i.e., the raw value
    // is within the confidence interval with probability of at least 0.1, then the hits count will
    // be at least 9552 with probability greater than 1 - 10^-6.
    assertThat(hits).isAtLeast(9552);
  }

  @Test
  public void computeConfidenceInterval_laplaceNoise_largeAlpha_satisfiesConfidenceLevel() {
    builder.noise(new LaplaceNoise()).delta(0.0);
    int rawCount = 14523;

    int hits = 0;
    for (int i = 0; i < 100000; i++) {
      Count count = builder.build();
      count.incrementBy(rawCount);
      count.computeResult();

      if (count.computeConfidenceInterval(/*alpha=*/ 0.9).lowerBound() <= rawCount
          && rawCount <= count.computeConfidenceInterval(/*alpha=*/ 0.9).upperBound()) {
        hits++;
      }
    }
    // Assuming that the true alpha of the confidence interval mechanism is 0.9, i.e., the raw value
    // is within the confidence interval with probability of at least 0.1, then the hits count will
    // be at least 9552 with probability greater than 1 - 10^-6.
    assertThat(hits).isAtLeast(9552);
  }

  @Test
  public void computeConfidenceInterval_calledBeforeComputeResult_throwsException() {
    Count count = builder.build();
    assertThrows(
        IllegalStateException.class, () -> count.computeConfidenceInterval(ARBITRARY_ALPHA));
  }

  @Test
  public void computeConfidenceInterval_calledAfterSerialization_throwsException() {
    Count count = builder.build();
    count.getSerializableSummary();
    assertThrows(
        IllegalStateException.class, () -> count.computeConfidenceInterval(ARBITRARY_ALPHA));
  }
}

File: go/dpagg/count_confidence_interval_test.go
//
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package dpagg

import (
	"testing"

	"github.com/google/differential-privacy/go/v3/noise"
)

func getCount(t *testing.T, n noise.Noise) *Count {
	t.Helper()
	delta := arbitraryDelta
	if n == noise.Laplace() {
		delta = 0.0
	}
	c, err := NewCount(&CountOptions{
		Epsilon:                  arbitraryEpsilon,
		Delta:                    delta,
		Noise:                    n,
		MaxPartitionsContributed: arbitraryMaxPartitionsContributed})
	if err != nil {
		t.Fatalf("Couldn't get count with noise=%v: %v", n, err)
	}
	return c
}

// Tests that Count.ComputeConfidenceInterval() does not return negative intervals.
func TestCountComputeConfidenceInterval_ClampsNegativeSubinterval(t *testing.T) {
	for i := 0; i < 1000; i++ {
		count := getCount(t, noise.Gaussian())
		_, err := count.Result()
		if err != nil {
			t.Fatalf("Couldn't compute dp result: %v", err)
		}

		// Using a large alpha to get a small confidence interval. This increases the chance of both
		// the lower and the upper bound being clamped.
		confInt, err := count.ComputeConfidenceInterval(0.99)
		if err != nil {
			t.Fatalf("Couldn't compute confidence interval: %v", err)
		}
		if confInt.LowerBound < 0 || confInt.UpperBound < 0 {
			t.Errorf("Confidence interval=%+v should have non-negative lower and upper bounds", confInt)
		}

		// Using a small alpha to get a large confidence interval. This increases the chance of only
		// the the upper bound being clamped.
		confInt, err = count.ComputeConfidenceInterval(0.01)
		if err != nil {
			t.Fatalf("Couldn't compute confidence interval: %v", err)
		}
		if confInt.LowerBound < 0 || confInt.UpperBound < 0 {
			t.Errorf("Confidence interval=%+v should have non-negative lower and upper bounds", confInt)
		}
	}
}

// Tests that Count.ComputeConfidenceInterval() returns the same interval when called for the same alpha twice.
func TestCountComputeConfidenceInterval_ReturnsSameResultForSameAlpha(t *testing.T) {
	for _, tc := range []struct {
		n         noise.Noise
		trueCount int64
	}{
		{noise.Gaussian(), 0}, // Clamping possible
		{noise.Gaussian(), 100_000_000},
		{noise.Laplace(), 0}, // Clamping possible
		{noise.Laplace(), 100_000_000},
	} {
		count := getCount(t, tc.n)
		count.IncrementBy(tc.trueCount)
		_, err := count.Result()
		if err != nil {
			t.Fatalf("With %v, couldn't compute dp result: %v", tc.n, err)
		}

		confInt1, err := count.ComputeConfidenceInterval(arbitraryAlpha)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		confInt2, err := count.ComputeConfidenceInterval(arbitraryAlpha)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		if confInt1.LowerBound != confInt2.LowerBound || confInt1.UpperBound != confInt2.UpperBound {
			t.Errorf("With %v, expected confInt1=%+v and confInt2=%+v with the same alpha to be equal", tc.n, confInt1, confInt2)
		}
	}
}

// Tests that Count.ComputeConfidenceInterval()'s result for small alpha is contained in the result
// for large alpha.
func TestCountComputeConfidenceInterval_ResultForSmallAlphaContainedInResultForLargeAlpha(t *testing.T) {
	for _, tc := range []struct {
		n         noise.Noise
		trueCount int64
	}{
		{noise.Gaussian(), 0}, // Clamping possible
		{noise.Gaussian(), 100_000_000},
		{noise.Laplace(), 0}, // Clamping possible
		{noise.Laplace(), 100_000_000},
	} {
		count := getCount(t, tc.n)
		count.IncrementBy(tc.trueCount)
		_, err := count.Result()
		if err != nil {
			t.Fatalf("With %v, couldn't compute dp result: %v", tc.n, err)
		}

		smallAlphaConfInt, err := count.ComputeConfidenceInterval(arbitraryAlpha * 0.5)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		largeAlphaConfInt, err := count.ComputeConfidenceInterval(arbitraryAlpha)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		if smallAlphaConfInt.LowerBound > largeAlphaConfInt.LowerBound {
			t.Errorf("With %v, expected smallAlphaConfInt's (%+v) lower bound to be smaller than largeAlphaConfInt's (%v) lower bound", tc.n, smallAlphaConfInt, largeAlphaConfInt)
		}
		if smallAlphaConfInt.UpperBound < largeAlphaConfInt.UpperBound {
			t.Errorf("With %v, expected smallAlphaConfInt's (%+v) upper bound to be larger than largeAlphaConfInt's (%v) upper bound", tc.n, smallAlphaConfInt, largeAlphaConfInt)
		}
	}
}

// Tests that Count.ComputeConfidenceInterval() matches the underlying noise implementation's
// confidence interval with the same parameters.
func TestCountComputeConfidenceInterval_MatchesNoiseConfidenceInterval(t *testing.T) {
	for _, tc := range []struct {
		n     noise.Noise
		delta float64
	}{
		{noise.Gaussian(), arbitraryDelta},
		{noise.Laplace(), 0.0},
	} {
		lInf := int64(1) // Always 1 for Count.
		count := getCount(t, tc.n)
		err := count.IncrementBy(100_000_000) // incrementing by large number to prevent clamping
		if err != nil {
			t.Fatalf("With %v, couldn't increment count: %v", tc.n, err)
		}
		result, err := count.Result()
		if err != nil {
			t.Fatalf("With %v, couldn't compute dp result: %v", tc.n, err)
		}

		countConfInt, err := count.ComputeConfidenceInterval(arbitraryAlpha)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		noiseConfInt, err := tc.n.ComputeConfidenceIntervalInt64(result, arbitraryMaxPartitionsContributed, lInf, arbitraryEpsilon, tc.delta, arbitraryAlpha)
		if err != nil {
			t.Fatalf("With %v, couldn't compute confidence interval: %v", tc.n, err)
		}
		if countConfInt.LowerBound != noiseConfInt.LowerBound || countConfInt.UpperBound != noiseConfInt.UpperBound {
			t.Errorf("With %v, countConfInt (%+v) and noiseConfInt (%+v) to be equal", tc.n, countConfInt, noiseConfInt)
		}
	}
}

// Tests that Count.ComputeConfidenceInterval() satisfies the confidence level for a given alpha.
func TestCountComputeConfidenceInterval_SatisfiesConfidenceLevel(t *testing.T) {
	rawCount := int64(14523)
	for _, tc := range []struct {
		n        noise.Noise
		alpha    float64
		wantHits int
	}{
		// Assuming that the true alpha of the confidence interval mechanism is 0.1, i.e., the raw value
		// is within the confidence interval with probability of at least 0.9, then the hits count will
		// be at least 89546 with probability greater than 1 - 10⁻⁶.
		{noise.Gaussian(), 0.1, 89546},
		{noise.Laplace(), 0.1, 89546},
		// Assuming that the true alpha of the confidence interval mechanism is 0.9, i.e., the raw value
		// is within the confidence interval with probability of at least 0.1, then the hits count will
		// be at least 9552 with probability greater than 1 - 10⁻⁶.
		{noise.Gaussian(), 0.9, 9552},
		{noise.Laplace(), 0.9, 9552},
	} {
		hits := 0
		for i := 0; i < 100000; i++ {
			count := getCount(t, tc.n)
			err := count.IncrementBy(rawCount)
			if err != nil {
				t.Fatalf("With %v, couldn't increment count: %v", tc.n, err)
			}
			_, err = count.Result()
			if err != nil {
				t.Fatalf("With %v, couldn't compute dp result: %v", tc.n, err)
			}

			confInt, err := count.ComputeConfidenceInterval(tc.alpha)
			if err != nil {
				t.Fatalf("With noise=%v alpha=%f, couldn't compute confidence interval: %v", tc.n, tc.alpha, err)
			}

			if confInt.LowerBound <= float64(rawCount) && float64(rawCount) <= confInt.UpperBound {
				hits++
			}
		}
		if hits < tc.wantHits {
			t.Errorf("With noise=%v alpha=%f, got %d hits, i.e. raw output within the confidence interval, wanted at least %d", tc.n, tc.alpha, hits, tc.wantHits)
		}
	}
}

// Tests that Count.ComputeConfidenceInterval() returns errors correctly with different Count aggregation states.
func TestCountComputeConfidenceInterval_StateChecks(t *testing.T) {
	for _, tc := range []struct {
		state   aggregationState
		wantErr bool
	}{
		{resultReturned, false},
		{defaultState, true},
		{merged, true},
		{serialized, true},
	} {
		c := getNoiselessCount(t)
		c.state = tc.state

		if _, err := c.ComputeConfidenceInterval(0.1); (err != nil) != tc.wantErr {
			t.Errorf("ComputeConfidenceInterval: when state %v for err got %v, wantErr %t", tc.state, err, tc.wantErr)
		}
	}
}
